assets/src/models/vehiclesByRouteId.ts

```diff
@@ -0,0 +1,54 @@
+import { partition } from "../helpers/array"
+import { ByRouteId, RouteId, Vehicle, VehiclesForRoute } from "../skate"
+
+interface NextAndPreviousVehicle {
+  nextVehicle?: Vehicle
+  previousVehicle?: Vehicle
+}
+
+export const allVehicles = (
+  vehiclesForRoute: VehiclesForRoute | undefined
+): Vehicle[] => {
+  if (vehiclesForRoute === undefined) {
+    return []
+  }
+
+  return [
+    ...vehiclesForRoute.onRouteVehicles,
+    ...vehiclesForRoute.incomingVehicles,
+  ]
+}
+
+export const allVehiclesForRoute = (
+  vehiclesByRouteId: ByRouteId<VehiclesForRoute>,
+  routeId: RouteId
+): Vehicle[] => allVehicles(vehiclesByRouteId[routeId])
+
+/**
+ * Partition vehicles by direction
+ */
+export const byDirection = (vehicles: Vehicle[]): Vehicle[][] =>
+  partition(vehicles, (vehicle: Vehicle): boolean => vehicle.directionId === 0)
+
+export const nextAndPreviousVehicle = (
+  vehicles: Vehicle[],
+  currentVehicle: Vehicle
+): NextAndPreviousVehicle => {
+  const vehiclesGoingTheSameDirection = byDirection(vehicles)[
+    currentVehicle.directionId
+  ]
+
+  const nextVehicle = vehiclesGoingTheSameDirection.find(
+    vehicle =>
+      vehicle.previousVehicleId === currentVehicle.id && vehicle.isOnRoute
```

Comment:
if I'm reading this correctly, it looks like if the next/previous vehicle is off-route, there will be no next/previous vehicle. should we instead look for the next vehicle after the off-route vehicle?

Reply:
Yes, that's true. We don't want to show the next or previous vehicle if it's currently in the incoming box. I think if the next vehicle is in the incoming box, the one after that will be too, so I don't think we need to look further. Can you think of a case I'm overlooking?

Reply:
oh, i was confusing this with the vehicle being off-course. this seems correct when the vehicel is off-route, but do we still need to check for the vehicle being off-course?

Reply:
At least for now (https://app.asana.com/0/1112935048846093/1133115437342310 might change it later), we are just taking what Swiftly says was the previous vehicle. So if that is now off course I think that seems fine to still show it.